marlowe-runtime/test/Language/Marlowe/Runtime/Transaction/ConstraintsSpec.hs

```diff
@@ -0,0 +1,178 @@
+  , payToRoles = Map.empty
+  , marloweOutputConstraints = TxConstraints.MarloweOutputConstraintsNone
+  , merkleizedContinuationsConstraints = mempty
+  , signatureConstraints = Set.empty
+  , metadataConstraints = Map.empty
+  }
+
+spec :: Spec
+spec = do
+  Hspec.QuickCheck.prop "mustSpendToken" do
+    policyId <- Chain.PolicyId <$> byteStringGen
+    tokenName <- Chain.TokenName <$> byteStringGen
+
+    let
+      assetId :: Chain.AssetId
+      assetId = Chain.AssetId policyId tokenName
+
+      actual :: Constraints.TxConstraints 'Core.Api.V1
+      actual = Constraints.mustSpendRoleToken assetId
+
+      expected :: Constraints.TxConstraints 'Core.Api.V1
+      expected = emptyTxConstraints
+        { roleTokenConstraints = Constraints.SpendRoleTokens $ Set.singleton assetId }
+
+    pure $ actual `shouldBe` expected
+
+  Hspec.describe "adjustTxForMinUtxo" do
+
+    Hspec.QuickCheck.prop "The Marlowe output has NOT been adjusted" do
+      protocol <- hedgehog genProtocolParameters
+
+      marloweScriptHash <- hedgehog genScriptHash
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        getValueAtAddress :: Chain.Address -> [C.TxOut C.CtxTx C.BabbageEra] -> Maybe (C.TxOutValue C.BabbageEra)
+        getValueAtAddress targetAddress txos = getFirst . mconcat . map (First
+          . (\(C.TxOut addressInEra txOutValue _ _) ->
+              if fromCardanoAddressInEra C.BabbageEra addressInEra == targetAddress
+                then Just txOutValue
+                else Nothing))
+          $ txos
+
+      txBodyContent <- hedgehog $ genTxBodyContent C.BabbageEra
+
+      let
+        actual = (getValueAtAddress marloweAddress . C.txOuts)
+          <$> Constraints.adjustTxForMinUtxo protocol marloweAddress txBodyContent
+        expected :: Either (Constraints.ConstraintError 'Core.Api.V1) (Maybe (C.TxOutValue C.BabbageEra))
+          = Right $ getValueAtAddress marloweAddress $ C.txOuts txBodyContent
+
+      pure $ actual `shouldBe` expected
+
+    Hspec.QuickCheck.prop "All outputs satisfy Cardano API minimum UTxO requirements" do
+      marloweScriptHash <- hedgehog genScriptHash
+
+      let
+        scriptAddress hash = fromCardanoAddressAny
+          $ C.AddressShelley
+          $ C.makeShelleyAddress C.Mainnet (C.PaymentCredentialByScript hash) C.NoStakeAddress
+        marloweAddress = scriptAddress marloweScriptHash
+
+        valueMeetsMinimumReq :: C.TxOut C.CtxTx C.BabbageEra -> Bool
+        valueMeetsMinimumReq txout@(C.TxOut _ txOrigValue _ _) =
+          case C.calculateMinimumUTxO C.ShelleyBasedEraBabbage txout protocolTestnet of
+            Right minValueFromApi ->
+              (C.selectLovelace . C.txOutValueToValue $ txOrigValue)
+              >= (C.selectLovelace minValueFromApi)
+            Left _ -> False
```

Comment:
Let's throw an error here actually, this is not expected to happen, and should be reported as an exception, not a property falsification.

Reply:
Now gathering up the failing cases and reporting on them with `expectationFailure`